Lecture 5

sazro19 opened the pull request.

sazro19 committed:
Task 5.1: implemented Order things, Checkout page

sazro19 committed:
Task 5.2: implemented overview page

sazro19 committed:
Task 5.3: implemented Dos protection

sazro19 committed:
Added tests

sazro19 committed:
Added mockito tests

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/exception/ItemNotFoundException.java:
```diff
@@ -0,0 +1,16 @@
+    }
+
+    public ItemNotFoundException() {
+    }
```
add super() / super("Item not found " + id)

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/order/PaymentMethod.java:
```diff
@@ -0,0 +1,25 @@
+    CACHE("Cache"),
+    CREDIT_CARD("Credit card");
+
+    private String value;
```
make final

Alexandra-ES commented on src/main/java/com/es/phoneshop/web/CheckoutPageServlet.java:
```diff
@@ -0,0 +1,120 @@
+        Cart cart = cartService.getCart(request.getSession());
+        Order order = orderService.getOrder(cart);
+        if (order == null) {
+            response.sendRedirect(request.getContextPath() + "/products");
```
why redirect to PLP?

Alexandra-ES commented on src/main/java/com/es/phoneshop/web/CheckoutPageServlet.java:
```diff
@@ -0,0 +1,120 @@
+            errors.put(OrderDetailsAttributes.PAYMENT_METHOD.getValue(), "payment method is required");
+        } else {
+            order.setPaymentMethod(PaymentMethod.getEnum(paymentMethod));
+        }
```
[opt] you might want to move field validation logic outside of servlet

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/cart/DefaultCartService.java:
```diff
@@ -112,6 +112,17 @@ public void delete(Cart cart, Long productId) {
+        lock.writeLock().lock();
+        try {
+            cart.getItems().clear();
+            cart.setTotalQuantity(0);
```
reset total price also

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/order/PaymentMethod.java:
```diff
@@ -0,0 +1,25 @@
+package com.es.phoneshop.model.product.order;
+
+public enum PaymentMethod {
+    CACHE("Cache"),
```
CASH

Alexandra-ES commented on src/main/java/com/es/phoneshop/web/CheckoutPageServlet.java:
```diff
@@ -0,0 +1,121 @@
+        } else {
+            request.setAttribute(ERRORS_ATTRIBUTE, errors);
+            request.setAttribute(ORDER_ATTRIBUTE, order);
+            doGet(request, response);
```
just add payment methods to request and forward

Alexandra-ES commented on src/main/java/com/es/phoneshop/web/OrderOverviewPageServlet.java:
```diff
@@ -0,0 +1,37 @@
+import java.io.IOException;
+
+public class OrderOverviewPageServlet extends HttpServlet {
+    private OrderDao orderDao;
```
use OrderService here

sazro19 committed:
- refactored;
- fixed recently viewed;
- added phone example;

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/order/DefaultOrderService.java:
```diff
@@ -54,9 +58,14 @@ public Order getOrder(Cart cart) {
-        return Arrays.stream(PaymentMethod.values())
-                .map(PaymentMethod::getValue)
-                .collect(Collectors.toList());
+        lock.readLock().lock();
```
this method doesn't need lock logic

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/order/DefaultOrderService.java:
```diff
@@ -66,12 +75,94 @@ public void placeOrder(Order order) {
+    @Override
+    public void checkDeliveryAddress(Order order, String address, Map<String, String> errors) {
+        setOrderFieldOrErrors(address, OrderDetailsAttributes.getDeliveryAddress(), errors, order::setDeliveryAddress);
+    }
```
imho these methods are excessive

Alexandra-ES commented on src/main/java/com/es/phoneshop/web/OrderDetailsAttributes.java:
```diff
@@ -1,20 +1,34 @@
-    DELIVERY_DATE("deliveryDate"),
-    DELIVERY_ADDRESS("deliveryAddress"),
-    PAYMENT_METHOD("paymentMethod");
+public class OrderDetailsAttributes {
```
make this class non instantiable

sazro19 committed:
- removed locks;
- OrderDetailsAttributes is non instantiable;
- removed one-line methods from services;
